src/algorithm/broadphase/sweep_prune.rs

```diff
@@ -0,0 +1,343 @@
+
+use std::clone::Clone;
+use std::cmp::Ordering;
+
+use num::NumCast;
+
+use self::variance::{Variance, Variance2, Variance3};
+use super::HasBound;
+use prelude::*;
+
+/// Broad phase sweep and prune algorithm for 2D, see
+/// [SweepAndPrune](struct.SweepAndPrune.html) for more information.
+pub type SweepAndPrune2<S> = SweepAndPrune<Variance2<S>>;
+
+/// Broad phase sweep and prune algorithm for 3D, see
+/// [SweepAndPrune](struct.SweepAndPrune.html) for more information.
+pub type SweepAndPrune3<S> = SweepAndPrune<Variance3<S>>;
+
+/// Sweep and prune broad phase collision detection algorithm.
+///
+/// Will sort the bounding boxes of the collision world along some axis, and will then sweep the
+/// sorted list, and compare the bounds along the sweep axis, adding them to an active list when
+/// they are encountered, and removing them from the active list when the end extent is passed.
+///
+/// Any shape pairs found by the base algorithm, will then do a bounding box intersection test,
+/// before adding to the resulting pairs list.
+///
+/// # Type parameters:
+///
+/// - `V`: Variance type used for computing what axis to use on the next iteration. Should be either
+///        `Variance2` or `Variance3`.
+pub struct SweepAndPrune<V> {
+    sweep_axis: usize,
+    variance: V,
+}
+
+impl<V> SweepAndPrune<V>
+where
+    V: Variance,
+{
+    /// Create a new sweep and prune algorithm, will use the X axis as the first sweep axis
+    pub fn new() -> Self {
+        Self::new_impl(0)
+    }
+
+    /// Create a new sweep and prune algorithm, starting with the given axis as the first sweep axis
+    pub fn new_impl(sweep_axis: usize) -> Self {
```

Comment:
if it's an implementation detail (as name suggests), it shouldn't be public. Otherwise, we'll need a ~~bigger boat~~ better name